### DISCUSSION TITLE
PyInstaller uses `/usr/bin/sh` from real root instead of embedding one for `os.system`

### DISCUSSION THREAD

**Discussion by GamePlayer-8** [pyinstaller/pyinstaller#8258] [Help]
Good evening,

There's an issue when using `os.system` in the executable. When I've compiled the app on `glibc 2.35` and used on `glibc 2.38` and I've got `sh: symbol lookup error: sh: undefined symbol: rl_trim_arg_from_keyseq`.

`/usr/bin/sh` on my system is pointing at `/usr/bin/bash` and when I've symlinked the `/usr/bin/sh` to `/usr/bin/dash` the problem is gone.

The issue is why couldn't PyInstaller embed the `/usr/bin/sh` (`readlink -f /usr/bin/sh` exactly) into the binary to avoid missing `/usr/bin/sh` in the system?

**Comment by bwoodsend** [comment]
`sh` isn't just one file. It's requires the entirety of glibc and GNU's coreutils package. You really want to take the entire OS with you?

**Reply by GamePlayer-8** [reply]
> You really want to take the entire OS with you?

Not really but would be nice to have an option to specify, which `sh` should be used for `os.system`

**Reply by rokm** [reply]
> > You really want to take the entire OS with you?
> 
> Not really but would be nice to have an option to specify, which `sh` should be used for `os.system`

Unless you're specifying the full path to the target executable, it will be the one that is found in `PATH` (so even if you did bundle `sh` with the frozen application, you would still need to ensure that `sys._MEIPASS` is in the `PATH`).

**Comment by rokm** [comment]
Since you're trying to run an external program, you should take a look at https://pyinstaller.org/en/latest/common-issues-and-pitfalls.html#launching-external-programs-from-the-frozen-application and reset `LD_LIBRARY_PATH` before your `os.system` call - then you shouldn't be running into issues with missing symbols with bash, either.

**Reply by rokm** [reply]
> The issue is why couldn't PyInstaller embed the `/usr/bin/sh` (`readlink -f /usr/bin/sh` exactly) into the binary to avoid missing `/usr/bin/sh` in the system?

If you really wanted `sh` to be collected into frozen application, you would need to instruct PyInstaller to do so, using [`--add-data`](https://pyinstaller.org/en/stable/usage.html#cmdoption-add-data) or [`--add-binary`](https://pyinstaller.org/en/stable/usage.html#cmdoption-add-binary) or their spec file equivalents. But as Brénainn already pointed out, that is usually not a good idea.
